# Расчет расстояний до объектов
## звезды типа RR Lyr $-$ стандартные свечи

<a href="https://i1.wp.com/www.jimmynewland.com/wp/wp-content/uploads/2018/03/ngc3201-rrlyrae.gif?zoom=2&resize=740%2C416"><img src="https://i1.wp.com/www.jimmynewland.com/wp/wp-content/uploads/2018/03/ngc3201-rrlyrae.gif?zoom=2&resize=740%2C416" width="85%" /></a> 

*На картинке несколько кадров, заметны переменные звёзды*

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/2/28/Inverse_square_law.svg/2560px-Inverse_square_law.svg.png" width="50%" />

*Закон обратных квадратов*

\begin{equation*}
L=\sigma{T^4}(4\pi{R^2})\>[\mathbf{W}]
\end{equation*}

Данные о температуре обычно получают, пользуясь законом смещения Вина:

\begin{equation*}
\lambda_{peak}\cdot{T}=constant\>[\mathbf{m\cdot{K}}]
\end{equation*}

Текущая задача: определить расстояние до шарового скопления NGC 3201, пользуясь стандартными свечами.

<img src="https://i0.wp.com/www.jimmynewland.com/wp/wp-content/uploads/2019/07/ngc-3201.jpg" width="75%" />

Для нашего исследования возьмём одну из звезд типа RR Lyr: [NGC 3201 2216](http://simbad.u-strasbg.fr/simbad/sim-id?Ident=Cl*+NGC+3201+++++LS+++++358&NbIdent=1).
Также нам нужен калибровочный источник, находящийся близко к интересующей нас звезде и попадающий в поле зрения инструментов. Возьмём [эту](http://simbad.u-strasbg.fr/simbad/sim-id?Ident=Cl*+NGC+3201+++CWFD++3-106&NbIdent=1) звезду.
Обратите внимание, что в этой области нет иных ярких источников, а почти все звёзды являются частью скопления.
<a href="https://i1.wp.com/www.jimmynewland.com/wp/wp-content/uploads/2019/08/rr-lryae-zoom.gif"><img src="https://i1.wp.com/www.jimmynewland.com/wp/wp-content/uploads/2019/08/rr-lryae-zoom.gif" width="75%" /></a>
<a href="https://i1.wp.com/www.jimmynewland.com/wp/wp-content/uploads/2019/08/field_zoomed.png">
<img src="https://i1.wp.com/www.jimmynewland.com/wp/wp-content/uploads/2019/08/field_zoomed.png" width="75%" /></a>

Данные получены с помощью [Aladin Sky Atlas](https://aladin.u-strasbg.fr/).

На кадре видны следующие объекты:
<img src="https://i1.wp.com/www.jimmynewland.com/wp/wp-content/uploads/2019/08/sources_zoomed.png" width="75%" /></a>
* inCl* - источник - звезда, часть скопления.
* BlueStraggler - голубой страгглер: на [английском](http://astronomy.swin.edu.au/cosmos/b/blue+stragglers), на [русском](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D0%BB%D1%83%D0%B1%D1%8B%D0%B5_%D1%81%D1%82%D1%80%D0%B0%D0%B3%D0%B3%D0%BB%D0%B5%D1%80%D1%8B).
* RGB* - звезды, покинувшие главную последовательность и ушедшие на [ветвь красных гигантов](http://astronomy.swin.edu.au/cosmos/H/Horizontal+Branch+stars) на ГР диаграмме.
* RRLyr - переменные звезды, похожие на [RR Lyr](http://astronomy.swin.edu.au/cosmos/R/RR+Lyrae) с периодом меньше суток и известной звездной величиной.
* Star - звезды, не являющиеся частью скопления (возможно).
* X - источники жесткого излучения: пульсары (нейтронные звезды) или черные дыры.

Далее - часть, в которой неиронично надо что-то делать(

In [3]:
# Установка немного кривого пакета, который я нашел случайно
# Его использует и обслуживает какая-то группа, https://github.com/kbarbary/sep
# ! - знак для запуска команд терминала прямо в ноутбуке
!pip install sep
#!pip install numpy
#!pip install matplotlib
#!pip install astropy

Defaulting to user installation because normal site-packages is not writeable


In [4]:
# Импорт всего, что потребуется
import numpy as np
import sep
import math
# Штуки для работы с файлами .fits, распространённый формат для хранения графической астроинформации
from astropy.utils.data import download_file
from astropy.io import fits

import matplotlib.pyplot as plt
from matplotlib import rcParams
# % - специфические команды для jupyter. Colab просто игнорирует, в других случаях может что-нибудь сломать, но вы быстро об этом узнаете
%matplotlib inline
# Определение размера картинок
rcParams['figure.figsize'] = [8., 8.]

In [28]:
# Значение звездной величины для Cl* NGC 3201 CWFD 3-106, наша калибровочная звезда
# http://simbad.u-strasbg.fr/simbad/sim-id?Ident=Cl*+NGC+3201+CWFD+3-106
m_calibration = 14.81
# Значение абсолютной звездной величины для Cl* NGC 3201 LS 358, наша стандартная свеча, на основе некоторой статьи
# https://iopscience.iop.org/article/10.1086/344948
M_target = 0.58

Для определения расстояния до звезды (а значит, и до скопления) нужно вычислить видимую звёздную величину. Это наиболее трудоемкая подзадача.

In [1]:
"""
    Функция для определения расстояния (в пк удобнее всего) до объекта из абсолютной и видимой звездной величины
"""
def distance_modulus(m,M):
    return ...

In [2]:
"""
    Проверим функцию на данных о Солнце:
    m = -26.76 , M =4.81
    Результат получите в а.е., должно получиться что-то около 1
"""
print()

In [9]:
"""
    Функция для получения звездной величины из потока,
    информация о котором поступает от инструмента
    Подробнее о переводе величин:
    http://classic.sdss.org/dr7/algorithms/fluxcal.html
"""
def app_mag(flux):
    return -2.5*math.log10(flux/(25.11*10**8))

In [8]:
"""
    Проверка функции конвертации, на основе данных о калибровочной звезде
    https://iopscience.iop.org/article/10.1086/344948
    Должно получиться 14.81
"""
print(app_mag(2967.8))

14.81853020554703


Далее некоторые технические функции для работы с данными

In [3]:
"""
    Функция для отражения картинки по обоим осям. Одно из изображений,
    которые мы будем использовать, перевернуто
"""
def mirror_data(data):
    dataout = np.array(data)
    rows = len(data)
    cols = len(data[0])
    for r in range(rows):
        for c in range(cols):
            ...
    return dataout

In [10]:
"""
    Получение данных из файла в формате .fits
"""
def get_image_data(filename, mirror=False):

    file = filename
    image_file = download_file(file, cache=True )

    data = fits.getdata(image_file)
    
    if mirror==True:
        data = mirror_data(data)
    
    # Эта штука нужна, чтобы с нашими данными мог работать пакет sep
    # https://sep.readthedocs.io/en/v1.0.x/tutorial.html#Finally-a-brief-word-on-byte-order
    data = data.byteswap().newbyteorder()
    
    return data

In [12]:
"""
    Эта функция будет вычитать из сырой картинки фон
"""
def subtract_background(data):
    # Вычисление пространственной вариации фона на картинке
    bkg = sep.Background(data)
    
    # Создание массива для фона
    bkg_image = bkg.back()
    # bkg_image = np.array(bkg) # то же самое, но через numpy
    
    # оценка величины фона, вычитание
    bkg_rms = bkg.rms()
    data_sub = data - bkg
    
    return data_sub, bkg

Работа со слабыми объектами осложняется шумом, он замыливает детали, ухудшает точность. Пакет sep позволяет решить эту проблему и повысить информативность данных.

In [14]:
# Скачивание и подготовка данных
data = get_image_data('http://jimmynewland.com/astronomy/ngc-3201/ngc3201_1.fits')
# Разделение на фон и очищенный сигнал
data_sub, bkg = subtract_background(data)
bkg_image = bkg.back()

In [6]:
# Рисование и нормировка, нечищенные данные
m1, s1 = np.mean(data), np.std(data)
plt.imshow(data, interpolation='nearest', cmap='gray',vmin=m1-s1, vmax=m1+s1, origin='lower')

NameError: name 'np' is not defined

In [5]:
# Рисование, фон, нормировка не требуется
plt.imshow()

NameError: name 'plt' is not defined

In [4]:
# Рисование очищенных данных, нормировка нужна
plt.imshow()

NameError: name 'plt' is not defined

In [20]:
"""
    Эта функция принимает в себя чистые данные, фон, а также координаты
    области, где находится интересующая нас звезда.
    Функция нарисует эллипсы вокруг всех найденных источников в области и выдаст
    список всех выделенных источников
"""
def extract_sources(data_sub, bkg, x1, y1, x2, y2):
    # поиск штук
    objects = sep.extract(data_sub, 1.5, err=bkg.globalrms)
    
    from matplotlib.patches import Ellipse

    # рисование чистых данных
    fig, ax = plt.subplots()
    m, s = np.mean(data_sub), np.std(data_sub)
    im = ax.imshow(data_sub, interpolation='nearest', cmap='gray',
               vmin=m-s, vmax=m+s, origin='lower')
    plt.xlim(x1,x2)
    plt.ylim(y1,y2)

    # Рисование эллипсов
    for i in range(len(objects)):
        e = Ellipse(xy=(objects['x'][i], objects['y'][i]),
                width=6*objects['a'][i],
                height=6*objects['b'][i],
                angle=objects['theta'][i] * 180. / np.pi)
        e.set_facecolor('none')
        e.set_edgecolor('red')
        ax.add_artist(e)
    plt.show()
    
    return objects

In [21]:
"""
    Эта функция помогает найти координаты целевой звезды:
    будем искать ближайший к центру области источник.
    Функция возвращает координаты центра прямоугольника с заданными координатами
"""
def select_target(x1, y1, x2, y2):

    ...
    
    return cen_x, cen_y

In [22]:
"""
    Эта функция суммирует поток по круглой области на картинке (arr)
    Технически - просто сумма пикселей
"""
def sum_circle(arr, cx, cy, r):
    
    ...
    
    return ...

In [23]:
"""
    Функция потребляет список найденных источников, чистые данные, фон и координаты центра
    
    Далее вычисляется поток от ближайшего к центру источника
"""
def get_target_flux(objects, data_sub, bkg, cen_x, cen_y):
    target_flux = 0
    
    for i in range(len(objects)):
        curr_x = objects[i]['x']
        curr_y = objects[i]['y']
        
        if curr_x > cen_x-5 and curr_x < cen_x+5 and curr_y > cen_y-5 and curr_y < cen_y+5:
            print(data_sub)
            flux = sum_circle(data_sub, objects[i]['x'], objects[i]['y'],3.0)
            target_flux = round(int(flux))
    return target_flux

In [24]:
"""
    Функция находит в списке источников калибровочный источник и вычисляет поток от него
"""
def get_calib_flux(objects, data_sub, bkg, cen_x, cen_y):
    for i in range(len(objects)):
        curr_x = objects[i]['x']
        curr_y = objects[i]['y']
            
        if curr_x > cen_x-5 and curr_x < cen_x+5 and curr_y > cen_y-81-15 and curr_y < cen_y-81+15:
            flux = sum_circle(data_sub, objects[i]['x'], objects[i]['y'],3.0)
            calibration_flux = round(int(flux))
    return calibration_flux

In [25]:
"""
"""
def process_image(filename, x1=0, y1=0, x2=2048, y2=2048, mirror=False):
    # Получение данных
    data = get_image_data(filename, mirror)
    
    # Получение фона
    data_sub, bkg = subtract_background(data)
    
    # Поиск источников
    objects = extract_sources(data_sub, bkg, x1, y1, x2, y2)
    
    # Нахождение центра кадра
    cen_x, cen_y = select_target(x1, y1, x2, y2)
    
    # Нахождение потока от целевого источника
    target_flux = get_target_flux(objects, data_sub, bkg, cen_x, cen_y)
    
    # Нахождение потока от калибровочного источника
    calibration_flux = get_calib_flux(objects, data_sub, bkg, cen_x, cen_y)
    
    # Преобразование потока в звездные величины
    m_cal_ap = app_mag(calibration_flux)
    
    # Нормировка значений для калибровочного источника
    f_cal = m_cal_ap/m_calibration # Калибровочное соотношение
    m_cal = m_cal_ap/f_cal 
    
    # Преобразование потока целевого источника в абсолютною звездную величину
    m_tar_ap = app_mag(target_flux)
    
    # Калибровка яркости целевого источника
    m_tar = m_tar_ap/f_cal
        
    print("App. mag of target star m = : "+str(m_tar))
    print("App. mag of calib. star m = : "+str(m_cal))
    
    # Return the magnitude for the distance calculation
    return m_tar

In [26]:
# Список для значений звездной величины целевого источника
target_mags = []

Данные для работы process_image:<br>
filename='http://jimmynewland.com/astronomy/ngc-3201/ngc3201_1.fits'<br>
    x1=860<br>
    x2=1060<br>
    y1=668<br>
    y2=868<br>
filename='http://jimmynewland.com/astronomy/ngc-3201/ngc3201_2.fits'<br>
    x1=869<br>
    x2=1069<br>
    y1=673<br>
    y2=873<br>
filename='http://jimmynewland.com/astronomy/ngc-3201/ngc3201_3.fits'<br>
    x1=205<br>
    x2=435<br>
    y1=190<br>
    y2=420<br>
filename='http://jimmynewland.com/astronomy/ngc-3201/ngc3201_4.fits'<br>
    x1=262<br>
    x2=462<br>
    y1=195<br>
    y2=395<br>
filename='http://jimmynewland.com/astronomy/ngc-3201/ngc3201_5.fits'<br>
    x1=641<br>
    x2=841<br>
    y1=709<br>
    y2=909<br>
    mirror=True<br>

In [7]:
# Смотрим на список звёздных величин целевого источника на каждом кадре
print(target_mags)

NameError: name 'target_mags' is not defined

In [9]:
# Из всех значений имеет смысл выбрать самое яркое
brightest = ...
# Учтем межзвездное поглощение, в направлении наших источников оно составляет 0.25 звездных величин
m_target = ...
print(m_target)

Ellipsis


In [40]:
distance = distance_modulus(m_target,M_target)/1000
print('Distance to NGC 3201: '+str(distance)+' kpc')

Distance to NGC 3201: 4.8657759946838715 kpc


Верное значение расстояния до NGC 3201 - 4.9 кпк.
Дополнительно можно ознакомиться с [этим](https://www.jimmynewland.com/wp/astro/measuring-rr-lyrae-stars-in-ngc-3201/) материалом, там еще и презентация есть.